presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeMatchingChecker.java

```diff
@@ -0,0 +1,421 @@
+        {
+            node.getProbeSource().accept(this, context);
+            node.getIndexSource().accept(this, context);
+
+            return null;
+        }
+
+        @Override
+        public Void visitIndexSource(IndexSourceNode node, Void context)
+        {
+            return null;
+        }
+
+        @Override
+        public Void visitTableScan(TableScanNode node, Void context)
+        {
+            return null;
+        }
+
+        @Override
+        public Void visitValues(ValuesNode node, Void context)
+        {
+            return null;
+        }
+
+        @Override
+        public Void visitUnnest(UnnestNode node, Void context)
+        {
+            PlanNode source = node.getSource();
+            source.accept(this, context);
+
+            return null;
+        }
+
+        @Override
+        public Void visitRemoteSource(RemoteSourceNode node, Void context)
+        {
+            return null;
+        }
+
+        @Override
+        public Void visitExchange(ExchangeNode node, Void context)
+        {
+            for (int i = 0; i < node.getSources().size(); i++) {
+                PlanNode subplan = node.getSources().get(i);
+                subplan.accept(this, context); // visit child
+            }
+
+            return null;
+        }
+
+        @Override
+        public Void visitTableWriter(TableWriterNode node, Void context)
+        {
+            PlanNode source = node.getSource();
+            source.accept(this, context); // visit child
+
+            return null;
+        }
+
+        @Override
+        public Void visitDelete(DeleteNode node, Void context)
+        {
+            PlanNode source = node.getSource();
+            source.accept(this, context); // visit child
+
+            return null;
+        }
+
+        @Override
+        public Void visitMetadataDelete(MetadataDeleteNode node, Void context)
+        {
+            return null;
+        }
+
+        @Override
+        public Void visitTableFinish(TableFinishNode node, Void context)
+        {
+            node.getSource().accept(this, context); // visit child
+
+            return null;
+        }
+
+        @Override
+        public Void visitUnion(UnionNode node, Void context)
+        {
+            for (int i = 0; i < node.getSources().size(); i++) {
+                PlanNode subplan = node.getSources().get(i);
+                subplan.accept(this, context); // visit child
+            }
+            checkMappingTypeMatching(node.getSymbolMapping());
+
+            return null;
+        }
+
+        @Override
+        public Void visitEnforceSingleRow(EnforceSingleRowNode node, Void context)
+        {
+            node.getSource().accept(this, context); // visit child
+
+            return null;
+        }
+
+        private void checkMappingTypeMatching(ListMultimap<Symbol, Symbol> symbolMapping)
+        {
+            for (Symbol symbol : symbolMapping.keySet()) {
```

Comment:
maybe one symbol could be named `keySymbol` and the other `valueSymbol` instead `symbol` and `element`